anesthetic/utils.py

```diff
@@ -264,6 +264,46 @@ def iso_probability_contours_from_samples(pdf, contours=[0.68, 0.95],
     return c
 
 
+def iso_probability_1d(x, pdf, contours=[0.68, 0.95]):
```

Comment:
I'm not entirely satisfied with this implementation involving turning points, as I think it relies on certain assumed properties of what a given PDF will look like.

 Is there any reason why the same scheme (or indeed the same functions `iso_probability_contours` or `iso_probability_contours_from_samples`)  couldn't be used here?

Reply:
I am not sure I follow the method in iso_probability_contours but from the looks of it it uses a cdf to find the contours? 

This is slightly different to what I am proposing for the 1D posteriors I think. Interpolation of a cdf will give you contours around the mean value of the posterior or have I got that wrong? Rather than contours corresponding to where the pdf=0.68 ect.

The interpolation of the pdf divided at the turning points should work regardless of the number of turning points in the pdf (i.e. there is no shape assumption here).

Reply:
Can you explain why the method used in 2d contours would not work for 1d 'contours' as well?

Reply:
I'm not entirely sure I follow what is happening in the method to calculate 2d contours.

But I think that the method calculates a cdf from the pdf. Then you interpolate this at say 0.32 and 0.68 to get the corresponding x value which, for a 1d posterior, gives you *two* limits to fill between for the 68% confidence interval. Whereas the method I have given interpolates the pdf at 0.32 giving in the example here 4 limits. Slightly different definitions of a confidence region right?

Using a cdf based method for the same example problem I'm showing will give a graph like this.
![cdf_limits](https://user-images.githubusercontent.com/40355093/107277951-10188300-6a4d-11eb-9bfd-417731fab8b0.png)

Reply:
...or something similar at least.

Reply:
It calculates a CDF as a function of p (probability) not x (position), which I believe is what you have computed above.

To compute the crossing points with the CDF based approach, you would have to compute the probability at which e.g. the 68% contour occurs (using the CDF as a function of p), and then look for crossing points in the PDF with this value.

>Can you explain why the method used in 2d contours would not work for 1d 'contours' as well?

I think this is the key thing you need to think about. The CDF approach works for 2D, which arguably is substantially more challenging than 1D.

Reply:
Ahh okay, I am interpolating my cdf as a function of position if this is what you mean? I will have a think and see if I can make this work for 1D, should be a lot simpler than 2D I agree.

Reply:
@williamjameshandley I have the workings of a code to calculate these credibility intervals with a cdf. I haven't yet fitted it into `plot.py` but it relies on this function `iso_probability_contours` in `utils`. My only concern is that the results look very different to those in my example figure and the reasoning behind the method doesn't make much sense to me. I would of thought that the 68% probability interval would correspond to p=0.32, which is what my original approach gives, but this method gives something like p=0.70.

![cdf_method](https://user-images.githubusercontent.com/40355093/107349512-37f4fe80-6ac0-11eb-8b13-ac1e5cbfabcb.png)

The code (I will tidy this later):
          
          import numpy as np
          import matplotlib.pyplot as plt
          from anesthetic.utils import sample_compression_1d, iso_probability_contours
          from scipy.stats import gaussian_kde
          
          data = np.hstack([np.random.normal(0, 3, 1000), np.random.normal(10, 2, 2000)])
          
          x, w = sample_compression_1d(data, None, 1000)
          kde = gaussian_kde(x, weights=None)
          p = kde(x)
          p /= p.max()
          
          levels = iso_probability_contours(p, contours=[0.68, 0.95])
          print(levels)
          idx = np.where(p >= levels[0])[0]
          idx2 = np.where(p >= levels[1])[0]
          
          # Bit messy but I fill fix this when I add it into utils/plots
          split = []
          for i in range(1, len(idx)):
              if not np.isclose(idx[i] - idx[i-1], 1, rtol=1e-1):
                  split = idx[i-1]
                  split_far = idx[i]
          
          print(idx[0], split)
          plt.plot(x, p)
          plt.fill_between(x[idx[0]:split], p[idx[0]:split],
              color='blue', alpha=0.3, label='68%')
          plt.fill_between(
              x[split_far:idx[-1]], p[split_far:idx[-1]],
              color='blue', alpha=0.3)
          plt.fill_between(x[idx2], p[idx2], color='blue', alpha=0.9, label='95%')
          plt.savefig('cdf_method.png')
          plt.show()

Reply:
Ahh we need to be using iso_probabiliy_contours_from_samples to get this working properly by the looks.

About to push some changes to the repo which should fix this.